src/FwdState.cc

```diff
@@ -1173,9 +1217,12 @@ FwdState::dispatch()
 
 #if USE_OPENSSL
     if (request->flags.sslPeek) {
+        // we were just asked to peek at the server, and we did that
         CallJobHere1(17, 4, request->clientConnectionManager, ConnStateData,
                      ConnStateData::httpsPeeked, ConnStateData::PinnedIdleContext(serverConnection(), request));
         unregister(serverConn); // async call owns it now
+        flags.dont_retry = true; // we gave up forwarding control
+        flags.abandonEntry = true; // TODO: or should we abort it instead?
```

Comment:
The overlapping names of the existing StoreEntry::abandon()  and the new 'abandonEntry' flag may be confusing and raise (maybe stupid) questions like "why can't we immediately StoreEntry::abandon() instead of remembering this state with a specific flag"?

I would rather agree with this TODO - can we simply call entry->abort() here and in dispatch() instead of juggling with these flags? In this case, we would not need to add flags.abandonEntry, I guess.

Reply:
The overlap is somewhat intentional and, yes, it kind of indicates a problem with this code. I do not yet know whether we should use this small hack or fix this "properly". The answer depends on what a proper fix would require.

> can we simply call entry->abort() here and in dispatch() instead of juggling with these flags?

The answer to that (and TODO) question requires research. I do not know what happens when these SslBump-related entries are abort()ed, especially with respect to invokeHandlers() effects. See commit 1eb5d56 for a more detailed list of questions/concerns to research.

Reply:
We should have an empty entry here with the ENTRY_FWD_HDR_WAIT set. If so, invokeHandlers() will do nothing.

Reply:
This needs to be carefully researched/tested/investigated. We should not dismiss these concerns based on superficial analysis.

```C++
StoreEntry::abort()
{
...
    // allow the Store clients to be told about the problem
    EBIT_CLR(flags, ENTRY_FWD_HDR_WAIT);
...
    invokeHandlers();
}
```

Reply:
To investigate whether/how StoreEntry::invokeHandlers() is called (in the 'splice during step3' test) and what effects it produces, I ran the test with 3 Squids:
1. this PR:
No handlers are invoked because the list of clients is cleared in storeUnregster(). Here is the call stack:
```
StoreEntry::invokeHandlers() at store_client.cc:758 0x555555b1955c	
StoreEntry::abort() at store.cc:1,149 0x555555b0e480	
storeUnregister() at store_client.cc:742 0x555555b194ae	
Ssl::ServerBump::~ServerBump() at ServerBump.cc:54 0x555555cf7438	
ConnStateData::~ConnStateData() at client_side.cc:680 0x555555a482ac	
Http::One::Server::~Server() at Http1Server.h:26 0x555555c0a2fc	
Http::One::Server::~Server() at Http1Server.h:26 0x555555c0a368	
AsyncJob::callEnd() at AsyncJob.cc:146 0x555555c2dfe7	
```
2. master:
No handlers were invoked in my simple 1-client simple test because sc->_callback.pending() returned false.
```
StoreEntry::invokeHandlers() at store_client.cc:774 0x555555b21156	
StoreEntry::complete() at store.cc:1,071 0x555555b159e8	
FwdState::complete() at FwdState.cc:561 0x555555998935	
FwdState::connectedToPeer() at FwdState.cc:1,013 0x55555599c0f0	
```
3. patched PR:
I commented the `flags.abandonEntry = true;` in connectedToPeer() and added `entry->abort()`. The result was the same as in (2), with the difference that invokeHandlers() was initiated by abort().

So it looks like we use abort() in connectedToPeer() without problems. Note that I did not test another similar path, in  FwdState::dispatch().

Reply:
> 2. master: No handlers were invoked in my simple 1-client simple test because sc->_callback.pending() returned false.

> 3. patched PR: same as in (2), with the difference that invokeHandlers() was initiated by abort().

Hm... Please check why there is no callback in those cases. I would expect the Ssl::ServerBump object to be a registered/active StoreEntry client that expects to be notified about StoreEntry changes.

Reply:
> Ssl::ServerBump object to be a registered/active StoreEntry client 

The client is stored in ServerBump::sc, to be more precise. This object is indeed created, but its _callback field remains nil. It acquires its value in store_client::copy(), but we do not get there in this test.

Reply:
> ServerBump::sc ... is indeed created, but its _callback field remains nil. It acquires its value in store_client::copy()

That sounds like a Squid bug to me: The registered StoreEntry client probably wants to be notified about abort()s even if they happen before the first copy(). The client is waiting for stored data and should stop waiting if no data is coming. We may leave the arguably out-of-scope bug as is, but I do not want our code to _rely_ on it.

Please see if you can temporary hack the code so that abort() creates _callback (if needed) and calls store_client::callback() with an error, as expected. Feel free to add/call StoreEntry::invokeHandlers2() for that purpose. What happens then?

Reply:
>  temporary hack the code so that abort() creates _callback 

I don't think we need this: after examining ServerBump::sc I noticed that it is not passed to storeClientCopy() and other methods. It is documented as a " dummy client to prevent entry trimming", and it's callback is never assigned.

Reply:
> It is documented as a "dummy client to prevent entry trimming".

OK, that is useful to know, but I remain concerned that changes (perhaps even bug fixes) in the ServerBump code will enable [StoreEntry::abort()] notifications. Perhaps we can attack this from a different angle: Does that dummy client or surrounding ServerBump code has any methods to receive Store notifications? In other words, if we start passing `sc` to storeClientCopy() and friends, what function will invokeHandlers() call?

If there is no such function at all right now, then we can address my concerns by saying that it would be up to those who add that function to implement it correctly.

Reply:
> if we start passing sc to storeClientCopy() and friends, what function will invokeHandlers() call

Eventually it calls store_client::Callback::callback_handler, with void `*data, StoreIOBuffer result` parameters. There are several such callbacks, for example, clientReplyContext::SendMoreData(...), but none of them 'belongs' to the ServerBump's surroundings.

Reply:
I agree. Based on your research, I think we should address that TODO in this branch: Please replace flags.abandonEntry with StoreEntry::abort() calls and retest.

Reply:
Done(7948c20) and re-tested with Daft.